tests/test_browser.py

```diff
@@ -3238,6 +3238,9 @@ def test_webidl(self):
 
   @requires_sync_compilation
   def test_dynamic_link(self):
```

Comment:
I'd prefer to not modify the existing simple test, and add a new test (`test_dynamic_link_lsdo` or some other name) that tests the extra functionality added in this PR. That way a failure is more specific as to what has broken.

Reply:
I agree with this.   I think the only reason I didn't mention it is because the browser tests tend to be more monolithic and have a higher overhead per-test.  But in general I agree.

Reply:
Thanks for this comment. I initially considered adding another test in test_browser, but saw that browser.test_dynamic_link carries infrastructure to test in different kinds of in-browser environment, and I did not want to duplicate that. Anyway after you askd I've reworked the patch to drop any changes to that "simple" test and instead unroll the full equivalent of ldso tests from test_core, but in {wasm,asmjs}/{inside main,inside worker} browser environments.

This way it even uncovered an old concurrency bug in loadWebAssemblyModule (see 007608c2 "ldso: wasm: Don't assume that WebAssembly modules are loaded serially" for the fix), so thanks again for wondering here.

The changes to test_browser.py now look as follows:

```diff
diff --git a/tests/test_browser.py b/tests/test_browser.py
index 65507adab..9f0b4ceaa 100644
--- a/tests/test_browser.py
+++ b/tests/test_browser.py
@@ -23,6 +23,7 @@
 from tools import system_libs
 from tools.shared import PYTHON, EMCC, WINDOWS, FILE_PACKAGER, PIPE, SPIDERMONKEY_ENGINE, JS_ENGINES
 from tools.shared import try_delete, Building, run_process, run_js
+import test_core
 
 try:
   from http.server import BaseHTTPRequestHandler, HTTPServer
@@ -3299,6 +3300,56 @@ def test_dynamic_link(self):
     # same wasm side module works
     self.btest(self.in_dir('main.cpp'), '2', args=['-s', 'MAIN_MODULE=1', '-O2', '--pre-js', 'pre.js', '-s', 'WASM=1', '-s', 'EXPORT_ALL=1'])
 
+
+  # verify that dynamic linking works in all kinds of in-browser environments.
+  # don't mix different kinds in a single test.
+  def test_dylink_dso_needed_wasm(self):            self._test_dylink_dso_needed(1, 0)
+  def test_dylink_dso_needed_wasm_inworker(self):   self._test_dylink_dso_needed(1, 1)
+  def test_dylink_dso_needed_asmjs(self):           self._test_dylink_dso_needed(0, 0)
+  def test_dylink_dso_needed_asmjs_inworker(self):  self._test_dylink_dso_needed(0, 1)
+
+  @requires_sync_compilation
+  def _test_dylink_dso_needed(self, wasm, inworker):
+    # here we reuse test_core.test_dylink_dso_needed, but the code is run via browser.
+    print('\n# wasm=%d inworker=%d' % (wasm, inworker))
+    self.set_setting('WASM', wasm)
+    self.emcc_args += ['-O2']
+    t = test_core.binaryen2 if wasm else test_core.asm2
+
+    def do_run(src, expected_output):
+      # XXX there is no infrastructure (yet ?) to retrieve stdout from browser in tests.
+      # -> do the assert about expected output inside browser.
+      #
+      # we have to put the hook into post.js because in main it is too late
+      # (in main we won't be able to catch what static constructors inside
+      # linked dynlibs printed), and in pre.js it is too early (out is not yet
+      # setup by the shell).
+      with open('post.js', 'w') as f:
+        f.write(r'''
+          Module.realPrint = out;
+          out = function(x) {
+            if (!Module.printed) Module.printed = "";
+            Module.printed += x + '\n'; // out is passed str without last \n
+            Module.realPrint(x);
+          };
+        ''')
+      src += r'''
+        int main() {
+          _main();
+          EM_ASM({
+            var expected = %r;
+            assert(Module.printed === expected, ['stdout expected:', expected]);
+          });
+          REPORT_RESULT(0);
+        }
+      ''' % (expected_output,)
+      # --proxy-to-worker only on main
+      if inworker:
+        self.emcc_args += ['--proxy-to-worker']
+      self.btest(src, '0', args=self.get_emcc_args() + ['--post-js', 'post.js'])
+
+    t._test_dylink_dso_needed.__func__(self, do_run)
+
   @requires_graphics_hardware
   @requires_sync_compilation
   def test_dynamic_link_glemu(self):
```

Please see whole interdiff for full details.